src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/utils/validators/SessionTemplateVisitMoveValidator.kt

```diff
@@ -0,0 +1,212 @@
+  }
+
+  private fun hasAllCategoriesMatch(fromSessionTemplate: SessionTemplateDto, toSessionTemplate: SessionTemplateDto): Boolean {
+    val fromSessionCategories =
+      sessionTemplateUtil.getPermittedPrisonerCategoryTypes(fromSessionTemplate.prisonerCategoryGroups)
+    val toSessionCategories =
+      sessionTemplateUtil.getPermittedPrisonerCategoryTypes(toSessionTemplate.prisonerCategoryGroups)
+
+    return sessionCategoryMatcher.hasAllMatch(fromSessionCategories, toSessionCategories)
+  }
+
+  private fun hasAllIncentiveLevelsMatch(fromSessionTemplate: SessionTemplateDto, toSessionTemplate: SessionTemplateDto): Boolean {
+    val fromSessionIncentiveLevels =
+      sessionTemplateUtil.getPermittedIncentiveLevels(fromSessionTemplate.prisonerIncentiveLevelGroups)
+    val toSessionIncentiveLevels =
+      sessionTemplateUtil.getPermittedIncentiveLevels(toSessionTemplate.prisonerIncentiveLevelGroups)
+
+    return sessionIncentiveLevelMatcher.hasAllMatch(fromSessionIncentiveLevels, toSessionIncentiveLevels)
+  }
+
+  private fun isValidSessionDate(allVisitDates: Set<LocalDate>, toSessionTemplate: SessionTemplateDto): Boolean {
+    return if (allVisitDates.isNotEmpty()) {
+      val firstVisitDate = allVisitDates.stream().min(Comparator.naturalOrder()).get()
+      val lastVisitDate = allVisitDates.stream().max(Comparator.naturalOrder()).get()
+
+      (isDateCoveredInSession(firstVisitDate, toSessionTemplate) && isDateCoveredInSession(lastVisitDate, toSessionTemplate))
+    } else {
+      true
+    }
+  }
+
+  private fun isValidWeeklyFrequency(fromSessionTemplate: SessionTemplateDto, toSessionTemplate: SessionTemplateDto, allVisitDates: Set<LocalDate>): Boolean {
+    // check if all existing visits can be accommodated in the new session template
+    if (fromSessionTemplate.weeklyFrequency % toSessionTemplate.weeklyFrequency != 0) {
+      allVisitDates.forEach {
+        if (!sessionDatesUtil.isActiveForDate(it, toSessionTemplate)) {
+          return false
+        }
+      }
+    }
+
+    return true
+  }
+
+  private fun doesSessionExceedCapacity(
+    fromSessionTemplateVisitStats: SessionTemplateVisitStatsDto,
+    toSessionTemplateVisitStats: SessionTemplateVisitStatsDto,
+    allowedToSessionCapacity: SessionCapacityDto,
+  ): List<LocalDate> {
+    val exceededSessionCapacityDates = mutableListOf<LocalDate>()
+    val toSessionVisits = toSessionTemplateVisitStats.visitsByDate?.associateBy { it.visitDate }
+    if (toSessionVisits != null) {
+      fromSessionTemplateVisitStats.visitsByDate?.forEach {
+        val toSessionTemplateVisitsForDate = toSessionVisits[it.visitDate]
+        if (toSessionTemplateVisitsForDate != null) {
+          val totalCapacityForVisitDate = sessionTemplateUtil.getTotalCapacity(it.visitCounts, toSessionTemplateVisitsForDate.visitCounts)
+          if (sessionTemplateUtil.sessionCapacityComparator.compare(allowedToSessionCapacity, totalCapacityForVisitDate) < 0) {
```

Comment:
```
val totalCapacityForVisitDate = sessionTemplateUtil.getTotalCapacity(it.visitCounts, toSessionTemplateVisitsForDate.visitCounts)
if (sessionTemplateUtil.sessionCapacityComparator.compare(allowedToSessionCapacity, totalCapacityForVisitDate) < 0) {
```

could this be put in the sessionTemplateUtil : 
```
if (sessionTemplateUtil.isCapacityExceeded(it.visitCounts, 
     toSessionTemplateVisitsForDate.visitCounts,
     allowedToSessionCapacity)) {
```
Something like that

Reply:
Updated code to use + operator and added isCapacityExceeded to SessionTemplateUtil as suggested.